test/puppetlabs/puppetdb/generative/overridable_generators.clj

```diff
@@ -0,0 +1,135 @@
+(ns puppetlabs.puppetdb.generative.overridable-generators
+  (:refer-clojure :exclude [vector boolean double hash-map keys]
+                  :rename {map clojure-core-map})
+  (:require [clojure.test.check.generators :as gen]
+            [com.gfredericks.test.chuck.generators :as chuck-gen]))
+
+;;; Overrideable generators
+;;;
+;;; This namespace provides a layer on top of regular test.check generators
+;;; which provides two things:
+;;; 1. A keyword-based registry for generators, in the spirit of clojure.spec.
+;;; 2. An 'override' combinator to locally rebind generators by name. (again,
+;;;    similar to the overrides feature in clojure.spec/gen)
+;;;
+;;; Implementation-wise, this happens by wrapping each generator in a function
+;;; of the overrides map; the function is then responsible for passing the
+;;; overrides map to its children. This is implemented in a generic way via the
+;;; `lookup` and `lift` functions.
+;;;
+;;; Overridable versions of the most common generators are predefined; others
+;;; can be added using the `lift` function or by directly creating
+;;; a `(fn [overrides] ...)` which returns a generator.
+;;;
+;;; To get a regular test.check generator back, suitable for use with
+;;; `prop/for-all`, use `convert`.
+
+;;; Generator registry and overridable lookup
+(defonce generators (atom {}))
+
+(defmacro defgen [gen-key body]
+  `(do
+     (swap! generators assoc ~gen-key ~body)
+     nil))
+
+(defn lookup
+  "given an overridable generator (w/optional overrides), or the keyword name of
+  one, return a test.check generator. "
+  ([gen] (lookup gen {}))
+  ([gen overrides]
+   (let [gens (merge @generators overrides)]
+     (cond
+       (keyword? gen) (if-let [gen-val (get gens gen)]
+                        (recur gen-val overrides)
+                        (throw (ex-info (str "Generator " gen " not registered") {:generator-key gen})))
+       (gen/generator? gen) gen
+       :default (gen overrides)))))
```

Comment:
You mean `(gen gens)`? No, I don't think so. I mean, I guess we could, but there's no point. In this case, 'gen' is an overridable genearator, i.e. a function of overrides -> generator. Each overridable generator calls `lookup` on its arguments, passing down overrides. So eventually it's going to bottom out in this `lookup` function again, having either a regular generator or a keyword. 

Which is to say, if we passed down `gens`, we'd end up merging in `@generators` twice. 

I do think moving the `let` into the `(keyword?)` block would be better though.